File: entities/src/main/java/net/kemitix/naolo/entities/Veterinarian.java
package net.kemitix.naolo.entities;

import lombok.*;

import javax.persistence.*;
import java.util.List;

@EntityListeners({
        JPAActivityListener.class
})
@Entity
@NamedQuery(name = Veterinarian.FIND_ALL,
        query = "Select v From Veterinarian v Order By v.name")
@With
@Setter
@Getter
@ToString
@EqualsAndHashCode
@NoArgsConstructor(force = true)
@AllArgsConstructor
public class Veterinarian
        implements HasId {

    public static final String FIND_ALL = "Veterinarian.FindAll";

    @Id
    @GeneratedValue
    long id;
    String name;
    @ElementCollection(
            targetClass = VetSpecialisation.class,
            fetch = FetchType.EAGER)
    @CollectionTable(
            name = "vet_specialisation",
            joinColumns = @JoinColumn(name = "vet_id"))
    @Column(name = "specialisation_id")
    List<VetSpecialisation> specialisations;
}

File: entities/src/main/scala/net/kemitix/naolo/entities/Veterinarian.scala
package net.kemitix.naolo.entities

import java.util
import scala.collection.JavaConverters._

/**
  * A Veterinarian.
  */
object Veterinarian {

  /**
    * Create a new Veterinarian object.
    *
    * @param id              the Veterinarian ID
    * @param name            the Veterinarians Name
    * @param specialisations the Specialisations of the Veterinarian
    * @return a new Veterinarian
    */
    def create(
                id: Long,
                name: String,
                specialisations: util.Set[VetSpecialisation]
              ) =
      new Veterinarian(id, name, specialisations.asScala.toStream.toSet)

  /**
    * Constructor parsing ';' delimited VetSpecialisation values.
    *
    * @param id              the Veterinarian ID
    * @param name            the Veterinarians Name
    * @param specialisations the Specialisations of the Veterinarian as a ';' delimited String
    */
  def create(
              id: Long,
              name: String,
              specialisations: String
            ) =
    new Veterinarian(id, name, specStringToSet(specialisations))

  def create() = new Veterinarian()

  private def specStringToSet(specialisations: String) = {
    specialisations.split(";").filter(_.length > 0)
      .map(VetSpecialisation.valueOf)
      .toSet
  }

}

final class Veterinarian private (
                                   id: Long,
                                   name: String,
                                   specialisations: Set[VetSpecialisation]
                                 ) {

  def this(
            id: Long,
            name: String,
            specialisations: String) =
    this(id, name, Veterinarian.specStringToSet(specialisations))

  def this() = this(0, "", "")

  def getId: Long = id
  def getName: String = name
  def getSpecialisations: util.Set[VetSpecialisation] = specialisations.asJava
}